### PR TITLE
Create a `PasswordProvider` wrapper object

### PR DESCRIPTION
**Pull request by richvdh**
The idea here is to abstract out all the conditional code which tests which
methods a given password provider has, to provide a consistent interface.

~~Based on #8835.~~

### CODE REVIEW

File: tests/handlers/test_password_providers.py [hunk 1]
```diff
@@ -266,8 +266,9 @@ def test_no_local_user_fallback_ui_auth(self):
         # first delete should give a 401
         channel = self._delete_device(tok1, "dev2")
         self.assertEqual(channel.code, 401)
-        # there are no valid flows here!
-        self.assertEqual(channel.json_body["flows"], [])
+        # m.login.password UIA is permitted because the auth provider allows it,
+        # even though the localdb does not.
+        self.assertEqual(channel.json_body["flows"], ["m.login.password"])
```
Comment (richvdh) [inline tests/handlers/test_password_providers.py:271]: this is actually me fixing a bug I introduced in #8835 :/

**Review by clokep — APPROVED** [review]
Nice clean-up. 👍

File: tests/handlers/test_password_providers.py (hunk 1 above)
Comment (clokep) [reply tests/handlers/test_password_providers.py:271]: This test seems to be failing actually?

Comment (richvdh) [reply tests/handlers/test_password_providers.py]: that's because I'm a silly.

**Pull request merged**